logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java

```diff
@@ -156,12 +193,55 @@ private void internalSetCommitPosition(final long commitPosition) {
 
   @Override
   public ActorFuture<LogStreamRecordWriter> newLogStreamRecordWriter() {
-    return actor.call(() -> new LogStreamWriterImpl(partitionId, writeBuffer));
+    final var writerFuture = new CompletableActorFuture<LogStreamRecordWriter>();
+    actor.call(() -> createWriter(writerFuture, LogStreamWriterImpl::new));
+    return writerFuture;
   }
 
   @Override
   public ActorFuture<LogStreamBatchWriter> newLogStreamBatchWriter() {
-    return actor.call(() -> new LogStreamBatchWriterImpl(partitionId, writeBuffer));
+    final var writerFuture = new CompletableActorFuture<LogStreamBatchWriter>();
+    actor.call(() -> createWriter(writerFuture, LogStreamBatchWriterImpl::new));
+    return writerFuture;
+  }
+
+  private <T extends LogStreamWriter> void createWriter(
+      final CompletableActorFuture<T> writerFuture, final WriterCreator<T> creator) {
+    final var alreadyOpenWriters = openWriterCount.getAndIncrement();
```

Comment:
We're in an actor context, no? Why do we need an atomic integer?

Reply:
Yes. At the beginning I was not 100% sure how I want to solve it. Either access the count directly, but I decided to sync with the actor. I can replace the count if you want.